starlite_sessions/sessions.py

```diff
@@ -0,0 +1,184 @@
+            send: The ASGI send function.
+
+        Returns:
+            None
+        """
+        if not self.has_wrapped_middleware:
+            starlite_app = cast("Starlite", scope["app"])
+            auth_middleware = SessionAuthMiddleware(
+                app=self.app,
+                exclude=self.config.exclude,
+                retrieve_user_handler=cast("AsyncCallable[[Dict[str, Any]], Any]", self.config.retrieve_user_handler),  # type: ignore
+            )
+            exception_middleware = ExceptionHandlerMiddleware(
+                app=auth_middleware,
+                exception_handlers=starlite_app.exception_handlers or {},
+                debug=starlite_app.debug,
+            )
+            self.app = SessionMiddleware(app=exception_middleware, config=self.config)
+        await self.app(scope, receive, send)
+
+
+class SessionAuthMiddleware(AbstractAuthenticationMiddleware):
+    def __init__(
+        self,
+        app: "ASGIApp",
+        exclude: Optional[Union[str, List[str]]],
+        retrieve_user_handler: AsyncCallable[[Dict[str, Any]], Any],
+    ):
+        """This is an abstract AuthenticationMiddleware that allows users to
+        create their own AuthenticationMiddleware by extending it and
+        overriding the 'authenticate_request' method.
+
+        Args:
+            app: An ASGIApp, this value is the next ASGI handler to call in the middleware stack.
+            exclude: A pattern or list of patterns to skip in the authentication middleware.
+            retrieve_user_handler: A callable that receives the session dictionary after it has been decoded and returns
+                a 'user' value.
+        """
+        super().__init__(app=app, exclude=exclude)
+        self.retrieve_user_handler = retrieve_user_handler
+
+    async def authenticate_request(self, connection: "HTTPConnection") -> AuthenticationResult:
+        """Implementation of the authentication method specified by Starlite's.
+
+        [AbstractAuthenticationMiddleware][starlite.middleware.authentication.AbstractAuthenticationMiddleware].
+
+        Args:
+            connection: A Starlette 'HTTPConnection' instance.
+
+        Raises:
+            [NotAuthorizedException][starlite.exceptions.NotAuthorizedException]: if session data is empty or user
+                is not found.
+
+        Returns:
+            [AuthenticationResult][starlite.middleware.authentication.AuthenticationResult]
+        """
+        if not connection.session or connection.session is Empty:
+            # the assignment of 'Empty' forces the session middleware to clear session data.
+            connection.scope["session"] = Empty
+            raise NotAuthorizedException("no session data found")
+
+        user = await self.retrieve_user_handler(connection.session)
+
+        if not user:
+            connection.scope["session"] = Empty
+            raise NotAuthorizedException("no user correlating to session found")
```

Comment:
For good UX, when the session is empty, web apps redirect the user to login page instead of raising 403 because most users will not know what this error is about.